packages/core/react/use-assistant.ts

```diff
@@ -0,0 +1,107 @@
+import { useState } from 'react';
+import { processMessageStream } from '../shared/process-message-stream';
+import { Message } from '../shared/types';
+import { getStreamStringTypeAndValue } from '../shared/utils';
+
+export type AssistantStatus = 'in_progress' | 'awaiting_message';
+
+export function useAssistant_experimental({
+  api,
+  threadId: threadIdParam,
+}: {
+  api: string;
+  threadId?: string | undefined;
+}) {
+  const [messages, setMessages] = useState<Message[]>([]);
+  const [input, setInput] = useState('');
+  const [threadId, setThreadId] = useState<string | undefined>(undefined);
+  const [status, setStatus] = useState<AssistantStatus>('awaiting_message');
+  const [error, setError] = useState<unknown | undefined>(undefined);
+
+  const handleInputChange = (e: any) => {
+    setInput(e.target.value);
+  };
+
+  const submitMessage = async (e: any) => {
+    e.preventDefault();
+
+    if (input === '') {
+      return;
+    }
+
+    setStatus('in_progress');
+
+    setMessages(messages => [
+      ...messages,
+      { id: '', role: 'user', content: input },
+    ]);
+
+    setInput('');
+
+    const result = await fetch(api, {
+      method: 'POST',
+      headers: { 'Content-Type': 'application/json' },
+      body: JSON.stringify({
+        // always use user-provided threadId when available:
+        threadId: threadIdParam ?? threadId ?? null,
+        message: input,
+      }),
+    });
+
+    if (result.body == null) {
+      throw new Error('The response body is empty.');
+    }
+
+    await processMessageStream(result.body.getReader(), (message: string) => {
+      try {
+        const { type, value } = getStreamStringTypeAndValue(message);
+        const messageContent = value as any;
```

Comment:
Increasing type safety here would be nice. Ideally this would happen in `getStreamStringTypeAndValue`